File: HighFreq/class1/mergeIntervals.py
"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    """
    @param intervals: interval list.
    @return: A new interval list.
    """
    def merge(self, intervals):
        # write your code here
        
        intervals = sorted(intervals, key = lambda interval:interval.start)
        
        last = None 
        
        result = [] 
        
        for curt in intervals:
            
            if last == None or last.end < curt.start:
                
                result.append(curt)
                
                last = curt 
                
            else:
                
                last.end = max(last.end, curt.end)
                
        return intervals

File: interviews/ForUsAll/Java/MergeIntervals.java
import java.util.*;

class Interval {
      int start, end;
      Interval(int start, int end) {
           this.start = start;
           this.end = end;
      }
}

public class Solution {
    /**
     * @param intervals: interval list.
     * @return: A new interval list.
     */
    public List<Interval> merge(List<Interval> intervals) {
        // write your code here
        
        List<Interval> results = new ArrayList<Interval>();
        
        if (intervals == null || intervals.size() == 0) {
            return results;
        }
        
        intervals.sort(new Comparator<Interval>() {
            public int compare(Interval o1, Interval o2) {
                return o1.start - o2.start;
            }
        });
        
        Interval last = null; 
        
        for (Interval curt : intervals) {
            
            if (last == null || last.end < curt.start) {
                results.add(curt);
                last = curt;
            }
            
            else {
                last.end = Math.max(last.end, curt.end);
            }
        }
        
        return results;
    }
}